### DISCUSSION TITLE
OptimizeMolecule and OptimizeMoleculeConfs Argument?

### DISCUSSION THREAD

**Discussion by xiaoruiDong** [rdkit/rdkit#3819] [Q&A] [answered]
Hi everyone,

I am trying to use `Chem.rdForceFieldHelpers.OptimizeMolecule` and `OptimizeMoleculeConfs` in Python. I know there are things like `MMFFOptimizeMolecule`, but I hope to optimize molecules on a force field with constraints. However, I keep getting `ArgumentError` when using those functions. An example is provided below (constraint part neglected):

```
# Generate molecule
mol = Chem.MolFromSmiles('CCC[O]')
mol = Chem.AddHs(mol)
Chem.AllChem.EmbedMultipleConfs(mol, numConfs=10)

# Setup forcefield
mp = Chem.rdForceFieldHelpers.MMFFGetMoleculeProperties(mol)
ff = Chem.rdForceFieldHelpers.MMFFGetMoleculeForceField(mol, mp)

# Optimize
Chem.rdForceFieldHelpers.OptimizeMoleculeConfs(mol, ff, numThreads=1, maxIters=200)
```

The error I got is
```
ArgumentError: Python argument types in
    rdkit.Chem.rdForceFieldHelpers.OptimizeMoleculeConfs(Mol, ForceField)
did not match C++ signature:
    OptimizeMoleculeConfs(RDKit::ROMol {lvalue} mol, ForceFields::ForceField {lvalue} ff, int numThreads=1, int maxIters=200)
```

From my naive investigation, `mol`, `numThreads`, and `maxIters` should be good. Then, the assigned `ff` is problematic? I guess that this function is asking for `ForceField` but `ff` is actually `pyForceField`? Is that correct?

**Comment by greglandrum** [comment, marked as answer]
This looks like a bug in the OptimizeMoleculeConfs() functions. We'll take a look.

In the meantime, if you want to optimize all the conformers without providing the force field, you can do:
```
Chem.rdForceFieldHelpers.MMFFOptimizeMoleculeConfs(mol)
```
Or if you want to provide the force field (so that you can set constraints) you can loop over the conformers individually and provide the confId argument when you construct the force field.

**Reply by xiaoruiDong** [reply]
Thanks, I will post an issue for this!
